### PR TITLE
Move the renderer docs section upper [skip ci]

### PR DESCRIPTION
**Pull request by yuriy-fix**
Connected to https://github.com/vaadin/components-team-tasks/issues/400

Sentence about two ways of setting the content in the beginning was added intentionally to create more space before the first heading **Rendering**

---
This change is [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-overlay/113)

### CODE REVIEW

**Review by manolo — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/vaadin/vaadin-overlay/113)* status: 0 of 2 files reviewed, 4 unresolved discussions (waiting on @YuriyVaadin)

___
*[src/vaadin-overlay.html, line 143 at r1](https://reviewable.io/reviews/vaadin/vaadin-overlay/113#-LJXPsSW6sTtUWSAQVpg:-LJXPsSW6sTtUWSAQVph:b-nt06ss) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/1e0880fb2ad0ec88b2992e8ec29dcc97038bf1a5/src/vaadin-overlay.html#L143)):*
> ```HTML
>      * `<vaadin-overlay>` is an element for creating overlays. The content of the overlay can be
>      * populated in two ways: imperatively using renderer callback function and declarativele
>      * using Polymer Templates
> ```

nit: using -> by using  
___
*[src/vaadin-overlay.html, line 147 at r1](https://reviewable.io/reviews/vaadin/vaadin-overlay/113#-LJXQ29l1gZdYEDxaCoS:-LJXQ29l1gZdYEDxaCoT:bhd65xo) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/1e0880fb2ad0ec88b2992e8ec29dcc97038bf1a5/src/vaadin-overlay.html#L147)):*
> ```HTML
>      * ### Rendering
>      *
>      * By default, the overlay uses the content provided using the renderer callback function.
> ```

sounds better  `By default, the overlay uses the content provided by the renderer function`
___
*[src/vaadin-overlay.html, line 152 at r1](https://reviewable.io/reviews/vaadin/vaadin-overlay/113#-LJXT-ZT6CNgSbD3ppu0:-LJXT-ZT6CNgSbD3ppu1:b-icqnuj) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/1e0880fb2ad0ec88b2992e8ec29dcc97038bf1a5/src/vaadin-overlay.html#L152)):*
> ```HTML
>      * so users are able to check if there is already content in `root`. Generate content using
>      * properties from `model` object if needed, append it to `root` and control the state
>      * of the host element by accessing `owner`.
> ```

I would reword this paragraph in this way:

```
 * The renderer function provides `root`, `owner`, `model` arguments when applicable.
 * Generate DOM content by using `model` object properties if needed, append it to the `root`
 * element and control the state of the host element by accessing `owner`. Before generating new
 * content, users are able to check if there is already content in `root` for reusing it.
```
___
*[src/vaadin-overlay.html, line 167 at r1](https://reviewable.io/reviews/vaadin/vaadin-overlay/113#-LJXTAJi69IZ9rcVX7kA:-LJXTAJi69IZ9rcVX7kB:b-yatyg5) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/1e0880fb2ad0ec88b2992e8ec29dcc97038bf1a5/src/vaadin-overlay.html#L167)):*
> ```HTML
>      *
>      * **NOTE:** renderer function could be called several times, so it is a good practise to
>      * check the existing content.
> ```

Probably it's more instructive that in this demo we already apply this recommended practice, something like

```
 const overlay = document.querySelector('#overlay');
 overlay.renderer = root => {
  let content = root.firstElementChild;
  if (!content) {
     content = document.createElement('div');
     root.appendChild(content)
  }
  content.textContent = "Overlay content";
 };
```

But it might distract to the user reading doc at this point so I'm not 100% sure. Not a blocker
___
*[src/vaadin-overlay.html, line 172 at r1](https://reviewable.io/reviews/vaadin/vaadin-overlay/113#-LJXUN9H6UR7p3pzQbox:-LJXUN9H6UR7p3pzQboy:b-twvfec) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/1e0880fb2ad0ec88b2992e8ec29dcc97038bf1a5/src/vaadin-overlay.html#L172)):*
> ```HTML
>      *
>      * Alternatively, the content can be provided with Polymer Template.
>      * Overlay finds the child template and uses that in case renderer callback function is not provided.
> ```

the child template -> the first child template

**Review by web-padawan — CHANGES_REQUESTED** [review]
> Reviewed 1 of 2 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/vaadin/vaadin-overlay/113)* status: all files reviewed, 7 unresolved discussions (waiting on @YuriyVaadin)

___
*[src/vaadin-overlay.html, line 141 at r2](https://reviewable.io/reviews/vaadin/vaadin-overlay/113#-LJXSx_d9l5SMdXIxy7T:-LJXSx_d9l5SMdXIxy7U:bep4prk) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/82166cf5d4198e883aa2df14c6ca2445c0bb710a/src/vaadin-overlay.html#L141)):*
> ```HTML
>     /**
>      *
>      * `<vaadin-overlay>` is an element for creating overlays. The content of the overlay can be
> ```

consider using "a Web Component" instead of the "element" word
___
*[src/vaadin-overlay.html, line 143 at r2](https://reviewable.io/reviews/vaadin/vaadin-overlay/113#-LJXT4YW6e481LKU6MvJ:-LJXT4YX0xtJ_Xf7aIiT:b-ck1cbx) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/82166cf5d4198e883aa2df14c6ca2445c0bb710a/src/vaadin-overlay.html#L143)):*
> ```HTML
>      * `<vaadin-overlay>` is an element for creating overlays. The content of the overlay can be
>      * populated in two ways: imperatively using renderer callback function and declaratively
>      * using Polymer Templates.
> ```

BTW why do we capitalise Template here? Is it just for consistency with the demos? Maybe we could consider using Polymer's `
___
*[src/vaadin-overlay.html, line 150 at r2](https://reviewable.io/reviews/vaadin/vaadin-overlay/113#-LJXTbq_3jDqzMkxM5CR:-LJXTbq_3jDqzMkxM5CS:b-z4rdpq) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/82166cf5d4198e883aa2df14c6ca2445c0bb710a/src/vaadin-overlay.html#L150)):*
> ```HTML
>      *
>      * The renderer function provides `root`, `owner`, `model` arguments where applicable,
>      * so users are able to check if there is already content in `root`. Generate content using
> ```

maybe add clear statement that "root" refers to a DOM node (as mentioned during DX tests)
___
*[src/vaadin-overlay.html, line 166 at r2](https://reviewable.io/reviews/vaadin/vaadin-overlay/113#-LJXUKE9-E85GMExfDdf:-LJXUKE9-E85GMExfDdg:bm4e6lm) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/82166cf5d4198e883aa2df14c6ca2445c0bb710a/src/vaadin-overlay.html#L166)):*
> ```HTML
>      * **NOTE:** when the renderer property is defined, the `<template>` content is not used.
>      *
>      * **NOTE:** renderer function could be called several times, so it is a good practise to
> ```

typo: practice

**Review by yuriy-fix — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/vaadin/vaadin-overlay/113)* status: 1 of 2 files reviewed, 6 unresolved discussions (waiting on @web-padawan, @YuriyVaadin, and @manolo)

___
*[src/vaadin-overlay.html, line 147 at r1](https://reviewable.io/reviews/vaadin/vaadin-overlay/113#-LJXQ29l1gZdYEDxaCoS:-LJXVYwhAAySQxOvzoCe:bw3fet1) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/1e0880fb2ad0ec88b2992e8ec29dcc97038bf1a5/src/vaadin-overlay.html#L147)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

sounds better  `By default, the overlay uses the content provided by the renderer function`
</blockquote></details>

I was thinking about that and the wording `provided by` can make a wrong impression that `renderer` function should return something.
___
*[src/vaadin-overlay.html, line 152 at r1](https://reviewable.io/reviews/vaadin/vaadin-overlay/113#-LJXT-ZT6CNgSbD3ppu0:-LJXWGrBAz9ZfRKYPcKY:b-896fix) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/1e0880fb2ad0ec88b2992e8ec29dcc97038bf1a5/src/vaadin-overlay.html#L152)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

I would reword this paragraph in this way:

```
 * The renderer function provides `root`, `owner`, `model` arguments when applicable.
 * Generate DOM content by using `model` object properties if needed, append it to the `root`
 * element and control the state of the host element by accessing `owner`. Before generating new
 * content, users are able to check if there is already content in `root` for reusing it.
```
</blockquote></details>

Done.
___
*[src/vaadin-overlay.html, line 167 at r1](https://reviewable.io/reviews/vaadin/vaadin-overlay/113#-LJXTAJi69IZ9rcVX7kA:-LJXWVhG-KySjMl3oNXD:b-jq74hk) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/1e0880fb2ad0ec88b2992e8ec29dcc97038bf1a5/src/vaadin-overlay.html#L167)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

Probably it's more instructive that in this demo we already apply this recommended practice, something like

```
 const overlay = document.querySelector('#overlay');
 overlay.renderer = root => {
  let content = root.firstElementChild;
  if (!content) {
     content = document.createElement('div');
     root.appendChild(content)
  }
  content.textContent = "Overlay content";
 };
```

But it might distract to the user reading doc at this point so I'm not 100% sure. Not a blocker

</blockquote></details>

Has created with the exact same simple structure as in the template example provided below. For such use-cases with checking previous content users should go through the demos.
___
*[src/vaadin-overlay.html, line 172 at r1](https://reviewable.io/reviews/vaadin/vaadin-overlay/113#-LJXUN9H6UR7p3pzQbox:-LJXXA8g-9WveoBbaWOl:b-896fix) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/1e0880fb2ad0ec88b2992e8ec29dcc97038bf1a5/src/vaadin-overlay.html#L172)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

the child template -> the first child template
</blockquote></details>

Done.
___
*[src/vaadin-overlay.html, line 141 at r2](https://reviewable.io/reviews/vaadin/vaadin-overlay/113#-LJXSx_d9l5SMdXIxy7T:-LJXVSSh3OQJ7p-rKSGW:b-896fix) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/82166cf5d4198e883aa2df14c6ca2445c0bb710a/src/vaadin-overlay.html#L141)):*
<details><summary><i>Previously, web-padawan (Serhii Kulykov) wrote…</i></summary><blockquote>

consider using "a Web Component" instead of the "element" word
</blockquote></details>

Done.
___
*[src/vaadin-overlay.html, line 143 at r2](https://reviewable.io/reviews/vaadin/vaadin-overlay/113#-LJXT4YW6e481LKU6MvJ:-LJXVUC1Ed80gZckKXCl:b-h4c792) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/82166cf5d4198e883aa2df14c6ca2445c0bb710a/src/vaadin-overlay.html#L143)):*
<details><summary><i>Previously, web-padawan (Serhii Kulykov) wrote…</i></summary><blockquote>

BTW why do we capitalise Template here? Is it just for consistency with the demos? Maybe we could consider using Polymer's `
</blockquote></details>

Yes, it is just for consistency.
___
*[src/vaadin-overlay.html, line 150 at r2](https://reviewable.io/reviews/vaadin/vaadin-overlay/113#-LJXTbq_3jDqzMkxM5CR:-LJXVwwt9uVZLObWb05s:b8479u9) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/82166cf5d4198e883aa2df14c6ca2445c0bb710a/src/vaadin-overlay.html#L150)):*
<details><summary><i>Previously, web-padawan (Serhii Kulykov) wrote…</i></summary><blockquote>

maybe add clear statement that "root" refers to a DOM node (as mentioned during DX tests)
</blockquote></details>

Rephrased to `root element`. Exact wording about DOM node is added here: https://reviewable.io/reviews/vaadin/vaadin-overlay/112
___
*[src/vaadin-overlay.html, line 166 at r2](https://reviewable.io/reviews/vaadin/vaadin-overlay/113#-LJXUKE9-E85GMExfDdf:-LJXWLDBAuw8CIhCv6Mt:b-mf0s1b) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/82166cf5d4198e883aa2df14c6ca2445c0bb710a/src/vaadin-overlay.html#L166)):*
<details><summary><i>Previously, web-padawan (Serhii Kulykov) wrote…</i></summary><blockquote>

typo: practice
</blockquote></details>

American and British English issue :)
Note is removed as reusing of the content is mentioned above.

**Review by manolo — APPROVED** [review]
:lgtm:

> Reviewed 1 of 2 files at r1, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/vaadin/vaadin-overlay/113#-:-LJX_27l9FqoGVF0Ie_B:bnfp4nl)* status: all files reviewed, 3 unresolved discussions (waiting on @web-padawan)

___
*[src/vaadin-overlay.html, line 147 at r1](https://reviewable.io/reviews/vaadin/vaadin-overlay/113#-LJXQ29l1gZdYEDxaCoS:-LJXZmRz2SuempcHO_TT:b32wtd) ([raw file](https://github.com/vaadin/vaadin-overlay/blob/1e0880fb2ad0ec88b2992e8ec29dcc97038bf1a5/src/vaadin-overlay.html#L147)):*
<details><summary><i>Previously, YuriyVaadin (Yuriy Yevstihnyeyev) wrote…</i></summary><blockquote>

I was thinking about that and the wording `provided by` can make a wrong impression that `renderer` function should return something.
</blockquote></details>

ok

**Review by web-padawan — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/vaadin/vaadin-overlay/113#-:-LJXfl7sDo_xRzXEzLuf:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved

**Review by yuriy-fix — COMMENTED** [review]
Rebased on top of updated `renderer-preview`

> *[Reviewable](https://reviewable.io/reviews/vaadin/vaadin-overlay/113#-:-LJYM1o85S0wcLRva9-D:bmp1mx5)* status: 0 of 2 files reviewed, all discussions resolved (waiting on @manolo and @web-padawan)

**Review by limonte — APPROVED** [review]
